### PR TITLE
Boot image with the bundled initramfs support

### PR DESCRIPTION
**Pull request by lumag**
A version of #232 with `QCOM_BOOTIMG_BUNDLE_INITRAMFS` removed.

### CODE REVIEW

**Comment by bhupesh-sharma** [comment]
would it possible to add some comments / documentation on how to use this option in your PR?

**Review by bhupesh-sharma — COMMENTED** [review]
We will need to add initramfs-test-full-image.bb to these as well.

File: recipes-kernel/linux/linux-qcom-bootimg.inc [hunk 1]
```diff
@@ -79,11 +101,17 @@ python do_qcom_img_deploy() {
         if not os.path.exists(output_img):
             os.symlink(os.path.basename(output), output_img)
 
+        if initrd:
+            make_image("boot-initramfs-%s-%s.img", rootfs, initrd)
```
Comment (ndechesne) [inline recipes-kernel/linux/linux-qcom-bootimg.inc:105]: we probably want to have the INITRAMFS_IMAGE in the filename that we build, to avoid overriding the boot image when building multiple of them.

Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:105]: You can have only a single image in the `INITRAMFS_IMAGE`. Thus you can not build 'multiple of them'.

File: recipes-kernel/linux/linux-qcom-bootimg.inc [hunk 2]
```diff
@@ -1,5 +1,17 @@
 QIMG_DEPLOYDIR = "${WORKDIR}/qcom_deploy-${PN}"
 
+# Define INITRAMFS_IMAGE to create kernel+initramfs Android boot images in
+# addition to default boot images. For example add the following line to your
+# conf/local.conf:
```
Comment (lumag) [inline recipes-kernel/linux/linux-qcom-bootimg.inc:5]: @bhupesh-sharma here is the documentation (in the source code). Just define the `INITRAMFS_IMAGE`.

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 1 above)
Comment (ndechesne) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:105]: you can build several of them calling bitbake multiple times, e.g. changing the value INITRAMFS_IMAGE in between. it's a valid scenario.  for example we will want to build initram-kerneltest-image and initramfs-kerneltest-full-image very soon.

**Comment by lumag** [comment]
> We will need to add initramfs-test-full-image.bb to these as well.

I've changed `initramfs-test-image` in a way to disable usage of bbappends.

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 1 above)
Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:105]: @ndechesne the the images from the previous release would be removed from the deploy dir during preparation stage (as stale objects). Or are you thinking about boot images stashed somewhere?

Comment (ndechesne) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:105]: we will indeed publish the images on linaro.org. 
and i think it makes sense to identify what we built by filename, using one or another initramfs will really make a very different boot-initramfs-xxx.img file.

Comment (ndechesne) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:105]: well actually.. it is going to make things a bit more complicated in our release build scripts.. since we will need to run the build in 2 steps now, with and without INITRAMFS_IMAGE..

Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:105]: I see... I'll extend the code to include initramfs image name.
Actually no. The image w/o initramfs is still generated in the same pass. See the build script (or the output dir).

Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc]: Done

File: recipes-test/images/initramfs-test-image.bb [hunk 3]
```diff
@@ -24,3 +24,11 @@ PACKAGE_INSTALL += " \
...
     wpa-supplicant \
 "
+
+# We'd like to include extra packages provided by layers which we do not depend
+# on. This can be handled by .bbappends, but then image recipes including this
+# one would not get all these tools. So simulate dynamic bbappend here.
+#
+# Unfortunately BBFILES_DYNAMIC isn't honored by include directive, so include
+# files have to be put into the main layer's dir.
+include ${@ " ".join(map(lambda layer: "recipes-test/images/initramfs-test-image-" + layer + ".inc", d.getVar('BBFILE_COLLECTIONS').split()))}
```
Comment (ndechesne) [inline recipes-test/images/initramfs-test-image.bb:34]: this is a complicated hack... how about we do something simpler:

PACKAGE_INSTALL += " \
    ${@bb.utils.contains("BBFILE_COLLECTIONS", "meta-oe", "XXX YYY ZZZ", "", d)} \
    ${@bb.utils.contains("BBFILE_COLLECTIONS", "meta-networking", "XXX YYY ZZZ", "", d)} \
"

it is also a 're-implementation' of BBFILES_DYNAMIC, but isn't that simpler?

File: recipes-kernel/linux/linux-qcom-bootimg.inc [hunk 4]
```diff
@@ -32,9 +75,9 @@ python do_qcom_img_deploy() {
     else:
         bb.fatal("Unuspported ARCH %s" % arch)
 
-    if os.path.exists(output_img):
+    if os.path.lexists(output_img):
```
Comment (ndechesne) [inline recipes-kernel/linux/linux-qcom-bootimg.inc:78]: why this change? it might be a good change, but it's not related to $subject?

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 2 above)
Comment (bhupesh-sharma) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:5]: As @ndechesne mentioned on irc, just setting it on the command-line is not working. Have you tested that use-case?

File: recipes-kernel/linux/linux-qcom-bootimg.inc [hunk 5]
```diff
@@ -24,6 +46,27 @@ python do_qcom_img_deploy() {
     output_img =  os.path.join(qcom_deploy_dir, "boot-%s.img" % (kernel_link_name))
     output_sd_img =  os.path.join(qcom_deploy_dir, "boot-sd-%s.img" % (kernel_link_name))
 
+    def make_image(template, dtb_name, rootfs, initrd=definitrd):
+        baseinitrd = os.path.basename(initrd)
+        output = os.path.join(qcom_deploy_dir, template % {'dtb': dtb_name, 'kernel': kernel_image_name, 'initrd': baseinitrd})
+        output_link = os.path.join(qcom_deploy_dir, template % {'dtb': dtb_name, 'kernel': kernel_link_name, 'initrd': baseinitrd})
+        output_initramfs_link = os.path.join(qcom_deploy_dir, template % {'dtb': dtb_name, 'kernel': kernel_link_name, 'initrd': 'initramfs'})
```
Comment (ndechesne) [inline recipes-kernel/linux/linux-qcom-bootimg.inc:53]: why that?

File: recipes-kernel/linux/linux-qcom-bootimg.inc [hunk 6]
```diff
@@ -24,6 +46,27 @@ python do_qcom_img_deploy() {
     output_img =  os.path.join(qcom_deploy_dir, "boot-%s.img" % (kernel_link_name))
     output_sd_img =  os.path.join(qcom_deploy_dir, "boot-sd-%s.img" % (kernel_link_name))
 
+    def make_image(template, dtb_name, rootfs, initrd=definitrd):
+        baseinitrd = os.path.basename(initrd)
+        output = os.path.join(qcom_deploy_dir, template % {'dtb': dtb_name, 'kernel': kernel_image_name, 'initrd': baseinitrd})
```
Comment (ndechesne) [inline recipes-kernel/linux/linux-qcom-bootimg.inc:51]: it is a bit confusing since initrd is not 'used' in the template for the regular image.. it wasn't obvious at first sight. i am finding this code less and less obvious to read (and to maintain).. not sure what to do.

File: recipes-kernel/linux/linux-qcom-bootimg.inc [hunk 7]
```diff
@@ -47,21 +90,6 @@ python do_qcom_img_deploy() {
         def getVarDTB(name):
             return d.getVarFlag(name, dtb_name) or d.getVar(name)
 
-        def make_image(template, rootfs):
```
Comment (ndechesne) [inline recipes-kernel/linux/linux-qcom-bootimg.inc:50]: why did you move the function?

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 2 above)
Comment (ndechesne) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:5]: setting in .conf file works. it's setting on the command line which does not work.

Comment (bhupesh-sharma) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:5]: Yep, I tried it with .conf as well, but the query is whether that is the expected way. If yes, probably we can better document it.

Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:5]: Of course it is. Typically you never ever define variables in the shell environment. In fact the environment is sanitized before being fed into called programs. You have to whiltelist the variable before bitbake would accept it.

File: recipes-test/images/initramfs-test-image.bb (hunk 3 above)
Comment (lumag) [reply recipes-test/images/initramfs-test-image.bb:34]: Yep. I was just thinking about files. Will do it this way.

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 4 above)
Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:78]: Ah. I was fighting with broken symlinks not being updated at some point. I'll drop this (or move to a separate patch).

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 5 above)
Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:53]: To get `boot-initramfs-apq8016-sbc-dragonboard-410c.img` links in addition to `boot-initramfs-kerneltest-image-apq8016-sbc-dragonboard-410c.img`. Having explicit initramfs name is good, but I want also to have a simple/readable symlink.

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 6 above)
Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:51]: Yes. I'm also a bit unhappy about it. Let me see if I can improve it.

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 7 above)
Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc:50]: a part of my symlinks fight. Will move it back.

File: recipes-kernel/linux/linux-qcom-bootimg.inc (hunk 6 above)
Comment (lumag) [reply recipes-kernel/linux/linux-qcom-bootimg.inc]: @ndechesne cleaned things up a bit.

File: recipes-test/images/initramfs-test-image.bb (hunk 3 above)
Comment (lumag) [reply recipes-test/images/initramfs-test-image.bb]: Done

**Comment by ndechesne** [comment]
thanks for the clean up! It looks better now. I am going to merge this PR, since we need it right away.

However, I think we should discuss about what we have done here.. the more i look into it, the more I think something is wrong with the whole idea of building the boot image in the kernel recipe. It was convenient initially since the boot image only contains 'artifact' built by the kernel recipe, but as it gets more and more complicated, we should change that. 

we could move the boot image creation to a custom 'image' class , or probably even better investigate how we could use WIC since its purpose is exactly to build images from all the build artifacts, once all recipes are built.

We are building lots of different artifact:
* kernel and DTBs, modules
* images (ext4, ubifs, android sparse images which is not supported in OE now, cpio)
* boot images with different purpose: 1) to load kernel and jump to rootfs, 2) boot image with initramfs for testing (and no rootfs), 3) hopefully soon boot images with initramfs which pivot into an actual rootfs...

**Comment by lumag** [comment]
@ndechesne point noted. I will think whether we can improve this by shifting into initramfs recipes.

The major problem is that if one would like to build a kernel image w/o the initramfs, he'd still call bitbake virtual/kernel rather than some bitbake initramfs-dummy. Thus we'd probably end up with some code duplication.
